pkg/controller/externalnode/controller_test.go

```diff
@@ -0,0 +1,652 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package externalnode
+
+import (
+	"context"
+	"crypto/sha1" // #nosec G505: not used for security purposes
+	"encoding/hex"
+	"io"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	"antrea.io/antrea/pkg/client/clientset/versioned"
+	fakeclientset "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions"
+)
+
+var (
+	informerFactory crdinformers.SharedInformerFactory
+)
+
+func TestAddExternalNode(t *testing.T) {
+	for _, tc := range []struct {
+		externalNode     *v1alpha1.ExternalNode
+		expectedEntities []v1alpha2.ExternalEntity
+	}{
+		{
+			externalNode: &v1alpha1.ExternalNode{
+				ObjectMeta: metav1.ObjectMeta{Name: "vm1", Namespace: "ns1", Labels: map[string]string{"en": "vm1"}},
+				Spec: v1alpha1.ExternalNodeSpec{
+					Interfaces: []v1alpha1.NetworkInterface{{IPs: []string{"1.1.1.2"}}},
+				},
+			},
+			expectedEntities: []v1alpha2.ExternalEntity{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "vm1",
+						Namespace: "ns1",
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: "crd.antrea.io/v1alpha1",
+								Kind:       "ExternalNode",
+								Name:       "vm1",
+							},
+						},
+						Labels: map[string]string{"en": "vm1"},
+					},
+					Spec: v1alpha2.ExternalEntitySpec{
+						Endpoints: []v1alpha2.Endpoint{
+							{IP: "1.1.1.2"},
+						},
+						ExternalNode: "vm1",
+					},
+				},
+			},
+		},
+		{
+			externalNode: &v1alpha1.ExternalNode{
+				ObjectMeta: metav1.ObjectMeta{Name: "vm2", Namespace: "ns1", Labels: map[string]string{"en": "vm2"}},
+				Spec: v1alpha1.ExternalNodeSpec{
+					Interfaces: []v1alpha1.NetworkInterface{{Name: "ens192", IPs: []string{"1.1.1.3"}}},
+				},
+			},
+			expectedEntities: []v1alpha2.ExternalEntity{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      generateEntityName("vm2", "ens192"),
```

Comment:
Better to use the result of `generateEntityName("vm2", "ens192")` as the expected value so it can prevent unintended change of the function.